tests/robustness/validate/operations.go

```diff
@@ -52,45 +50,50 @@ func validateLinearizableOperationsAndVisualize(lg *zap.Logger, operations []por
 	}
 }
 
-func validateSerializableOperations(t *testing.T, lg *zap.Logger, operations []porcupine.Operation, persistedRequests []model.EtcdRequest) {
+func validateSerializableOperations(lg *zap.Logger, operations []porcupine.Operation, replay *model.EtcdReplay) (lastErr error) {
 	lg.Info("Validating serializable operations")
-	staleReads := filterSerializableReads(operations)
-	if len(staleReads) == 0 {
-		return
-	}
-	sort.Slice(staleReads, func(i, j int) bool {
-		return staleReads[i].Input.(model.EtcdRequest).Range.Revision < staleReads[j].Input.(model.EtcdRequest).Range.Revision
-	})
-	replay := model.NewReplay(persistedRequests)
-	for _, read := range staleReads {
+	for _, read := range operations {
 		request := read.Input.(model.EtcdRequest)
 		response := read.Output.(model.MaybeEtcdResponse)
-		validateSerializableOperation(t, replay, request, response)
+		err := validateSerializableRead(lg, replay, request, response)
+		if err != nil {
+			lastErr = err
+		}
 	}
+	return lastErr
 }
 
-func filterSerializableReads(operations []porcupine.Operation) []porcupine.Operation {
+func filterSerializableOperations(clients []report.ClientReport) []porcupine.Operation {
 	resp := []porcupine.Operation{}
-	for _, op := range operations {
-		request := op.Input.(model.EtcdRequest)
-		if request.Type == model.Range && request.Range.Revision != 0 {
-			resp = append(resp, op)
+	for _, client := range clients {
+		for _, op := range client.KeyValue {
+			request := op.Input.(model.EtcdRequest)
+			if request.Type == model.Range && request.Range.Revision != 0 {
+				resp = append(resp, op)
+			}
 		}
 	}
 	return resp
 }
 
-func validateSerializableOperation(t *testing.T, replay *model.EtcdReplay, request model.EtcdRequest, response model.MaybeEtcdResponse) {
+func validateSerializableRead(lg *zap.Logger, replay *model.EtcdReplay, request model.EtcdRequest, response model.MaybeEtcdResponse) error {
 	if response.PartialResponse || response.Error != "" {
-		return
+		return nil
 	}
 	state, err := replay.StateForRevision(request.Range.Revision)
 	if err != nil {
-		t.Fatal(err)
+		if response.Error == model.ErrEtcdFutureRev.Error() {
+			return nil
+		}
+		lg.Error("Failed validating serializable operation", zap.Any("request", request), zap.Any("response", response))
+		return fmt.Errorf("request about a future rev with response")
 	}
 
 	_, expectResp := state.Step(request)
-	if !reflect.DeepEqual(response.EtcdResponse.Range, expectResp.Range) {
-		t.Errorf("Invalid serializable response, diff: %s", cmp.Diff(response.EtcdResponse.Range, expectResp.Range))
+
+	if diff := cmp.Diff(response.EtcdResponse.Range, expectResp.Range); diff != "" {
+		lg.Error("Failed validating serializable operation", zap.Any("request", request), zap.String("diff", diff))
+		return fmt.Errorf("response didn't match expected")
```

Comment:
make the error a constant so it is easier to check in tests.

Reply:
That's not a huge priority. Feel free to open a followup PR.